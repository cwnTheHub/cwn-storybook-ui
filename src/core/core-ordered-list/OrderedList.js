import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import OrderedItem, { StyledOrderedItem } from "./OrderedItem/OrderedItem";
import Box from "../core-box/Box";
import { componentWithName } from "../../util-prop-types";
import { list } from "../../shared-styles";
import { safeRest } from "../../util-helpers";

const listStyleType = {
  upperAlpha: "upper-alpha",
  lowerAlpha: "lower-alpha",
  decimal: "decimal",
};

export const StyledOrderedList = styled(({ size, listStyle, ...rest }) => (
  <Box {...rest} />
))(({ listStyle }) => ({
  paddingLeft: "3rem",
  listStyleType: listStyleType[listStyle],
  ...list.base,
}));

const OrderedList = ({ listStyle, size, children, ...rest }) => (
  <StyledOrderedList
    {...safeRest(rest)}
    tag="ol"
    between={2}
    listStyle={listStyle}
  >
    {React.Children.toArray(children)
      .filter((child) => child)
      .map((child) => React.cloneElement(child, { size }))}
  </StyledOrderedList>
);

OrderedList.propTypes = {
  listStyle: PropTypes.oneOf(["decimal", "upperAlpha", "lowerAlpha"]),
  size: PropTypes.oneOf(["small", "medium", "large"]),
  children: componentWithName("OrderedItem").isRequired,
};

OrderedList.defaultProps = {
  listStyle: "decimal",
  size: "medium",
};

OrderedList.Item = OrderedItem;

export default OrderedList;

export { StyledOrderedItem };
