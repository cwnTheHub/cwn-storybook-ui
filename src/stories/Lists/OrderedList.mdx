import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Story,
} from "@storybook/blocks";

import * as OrderedListStories from "./OrderedList.stories";

<Meta of={OrderedListStories} />
<Title />
`` import {OrderedList} from '@nds_core/nds' ``

### PROPS & METHODS

<Controls />

_Note: An `OrderedList` should not be used adjacent to an `UnorderedList`._

<Canvas of={OrderedListStories.Decimal}/>