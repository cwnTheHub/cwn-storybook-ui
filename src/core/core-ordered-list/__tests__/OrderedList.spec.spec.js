import React from "react";
import { shallow, render, mount } from "enzyme";

import OrderedList from "../OrderedList";

describe("<OrderedList />", () => {
  const doRender = (overrides = {}) =>
    render(
      <OrderedList {...overrides}>
        <OrderedList.Item>Lorem ipsum</OrderedList.Item>
        {undefined && <OrderedList.Item>Lorem ipsum</OrderedList.Item>}
        <OrderedList.Item>Dolor sit amet</OrderedList.Item>
      </OrderedList>
    );

  const doShallow = (overrides = {}) =>
    shallow(
      <OrderedList {...overrides}>
        <OrderedList.Item>Lorem ipsum</OrderedList.Item>
        {undefined && <OrderedList.Item>Lorem ipsum</OrderedList.Item>}
        <OrderedList.Item>Dolor sit amet</OrderedList.Item>
      </OrderedList>
    );

  const doMount = (overrides = {}) =>
    mount(
      <OrderedList {...overrides}>
        <OrderedList.Item>Lorem ipsum</OrderedList.Item>
        {undefined && <OrderedList.Item>Lorem ipsum</OrderedList.Item>}
        <OrderedList.Item>Dolor sit amet</OrderedList.Item>
      </OrderedList>
    );

  it("renders", () => {
    const orderedList = doRender();

    expect(orderedList).toMatchSnapshot();
  });

  it("OrderList renders an HTML ol tag", () => {
    const orderedList = doMount();

    expect(orderedList.find("ol")).toBeTruthy();
  });

  it("OrderList.Item renders an HTML li tag", () => {
    const orderedListItem = mount(
      <OrderedList.Item>a list item</OrderedList.Item>
    );

    expect(orderedListItem.find("li")).toBeTruthy();
  });

  it("can have a list style", () => {
    let orderedList = doRender({ listStyle: undefined });
    expect(orderedList).toMatchSnapshot();

    orderedList = doRender({ listStyle: "upperAlpha" });
    expect(orderedList).toMatchSnapshot();
  });

  it("passes additional attributes to ol element", () => {
    const orderedList = doShallow({ id: "the-list", reversed: "true" });

    expect(orderedList.props().id).toEqual("the-list");
    expect(orderedList.props()["reversed"]).toEqual("true");
  });

  it("does not allow custom CSS", () => {
    const orderedList = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(orderedList.props()["className"]).not.toEqual("my-custom-class");
    expect(orderedList.props().style).toBe(undefined);
  });
});
