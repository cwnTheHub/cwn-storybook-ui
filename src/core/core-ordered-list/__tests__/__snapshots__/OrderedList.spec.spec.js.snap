// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderedList /> can have a list style 1`] = `
<ol
  class="sc-eqUAAy gxGxiH sc-fqkvVR guxdMx sc-fqkvVR guxdMx"
>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Lorem ipsum
    </span>
  </li>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Dolor sit amet
    </span>
  </li>
</ol>
`;

exports[`<OrderedList /> can have a list style 2`] = `
<ol
  class="sc-eqUAAy gxGxiH sc-fqkvVR kcbKFd sc-fqkvVR kcbKFd"
>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Lorem ipsum
    </span>
  </li>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Dolor sit amet
    </span>
  </li>
</ol>
`;

exports[`<OrderedList /> renders 1`] = `
<ol
  class="sc-eqUAAy gxGxiH sc-fqkvVR guxdMx sc-fqkvVR guxdMx"
>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Lorem ipsum
    </span>
  </li>
  <li
    class="sc-aXZVg jUpfkX"
  >
    <span
      class="sc-gEvEer inVZch"
    >
      Dolor sit amet
    </span>
  </li>
</ol>
`;
